src/devices/rpLidar2/sdk_original/sdk/src/arch/win32/net_serial.cpp

```diff
@@ -0,0 +1,358 @@
+        else
+            r_len = 0;
+    }
+
+    return r_len;
+}
+
+void raw_serial::flush( _u32 flags)
+{
+    PurgeComm(_serial_handle, PURGE_TXABORT | PURGE_RXABORT | PURGE_TXCLEAR | PURGE_RXCLEAR );
+}
+
+int raw_serial::waitforsent(_u32 timeout, size_t * returned_size)
+{
+    if (!isOpened() ) return ANS_DEV_ERR;
+    DWORD w_len = 0;
+    _word_size_t ans =0;
+
+    if (WaitForSingleObject(_wo.hEvent, timeout) == WAIT_TIMEOUT)
+    {
+        ans = ANS_TIMEOUT;
+        goto _final;
+    }
+    if(!GetOverlappedResult(_serial_handle, &_wo, &w_len, FALSE))
+    {
+        ans = ANS_DEV_ERR;
+    }
+_final:
+    if (returned_size) *returned_size = w_len;
+    return ans;
+}
+
+int raw_serial::waitforrecv(_u32 timeout, size_t * returned_size)
+{
+    if (!isOpened() ) return -1;
+    DWORD r_len = 0;
+    _word_size_t ans =0;
+
+    if (WaitForSingleObject(_ro.hEvent, timeout) == WAIT_TIMEOUT)
+    {
+        ans = ANS_TIMEOUT;
+    }
+    if(!GetOverlappedResult(_serial_handle, &_ro, &r_len, FALSE))
+    {
+        ans = ANS_DEV_ERR;
+    }
+    if (returned_size) *returned_size = r_len;
+    return ans;
+}
+
+int raw_serial::waitfordata(size_t data_count, _u32 timeout, size_t * returned_size)
+{
+    COMSTAT  stat;
+    DWORD error;
+    DWORD msk,length;
+    size_t dummy_length;
+
+    if (returned_size==NULL) returned_size=(size_t *)&dummy_length;
+
+    
+    if ( isOpened()) {
+        size_t rxqueue_remaining =  rxqueue_count();
+        if (rxqueue_remaining >= data_count) {
+            *returned_size = rxqueue_remaining;
+            return 0;
+        }
+    }
+
+    while ( isOpened() )
+    {
+        msk = 0;
+        SetCommMask(_serial_handle, EV_RXCHAR | EV_ERR );
+        if(!WaitCommEvent(_serial_handle, &msk, &_wait_o))
+        {
+            if(GetLastError() == ERROR_IO_PENDING)
+            {
+                if (WaitForSingleObject(_wait_o.hEvent, timeout) == WAIT_TIMEOUT)
+                {
+                    *returned_size =0;
+                    return ANS_TIMEOUT;
+                }
+
+                GetOverlappedResult(_serial_handle, &_wait_o, &length, TRUE);
+
+                ::ResetEvent(_wait_o.hEvent);
+            }else
+            {
+                ClearCommError(_serial_handle, &error, &stat);
+                 *returned_size = stat.cbInQue;
+                return ANS_DEV_ERR;
+            }
+        }
+
+        if(msk & EV_ERR){
+            // FIXME: may cause problem here
+            ClearCommError(_serial_handle, &error, &stat);
+        }
+
+        if(msk & EV_RXCHAR){
+            ClearCommError(_serial_handle, &error, &stat);
+            if(stat.cbInQue >= data_count)
+            {
+                *returned_size = stat.cbInQue;
+                return 0;
+            }
+        }
+    }
+    *returned_size=0;
+    return ANS_DEV_ERR;
+}
+
+size_t raw_serial::rxqueue_count()
+{
+    if  ( !isOpened() ) return 0;
+    COMSTAT  com_stat;
+    DWORD error;
+    DWORD r_len = 0;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'r_len' is assigned a value that is never used.](https://app.codacy.com/manual/robotology/yarp/pullRequest?prid=5162834)